File: compiler/bench/bstfind.fs
\ expected: 99999
\ BST find 100K times, sum of depths searched

100000 constant N
create left N cells allot
create right N cells allot
create key N cells allot
variable root
variable ncount

: insert ( val -- )
  ncount @ 0= if
    dup key 0 cells + !
    -1 left 0 cells + !
    -1 right 0 cells + !
    1 ncount !
    drop exit
  then
  0
  begin
    over key over cells + @
    2 pick < if
      dup cells right + @
      dup 0< if
        drop
        ncount @ over cells right + !
        2 pick ncount @ cells key + !
        -1 ncount @ cells left + !
        -1 ncount @ cells right + !
        1 ncount +!
        2drop exit
      then
      nip
    else
      dup cells left + @
      dup 0< if
        drop
        ncount @ over cells left + !
        2 pick ncount @ cells key + !
        -1 ncount @ cells left + !
        -1 ncount @ cells right + !
        1 ncount +!
        2drop exit
      then
      nip
    then
  again ;

: find-depth ( val -- depth )
  0 0  \ depth node
  begin
    over key over cells + @
    3 pick = if 2drop nip exit then
    over key over cells + @
    3 pick < if
      dup cells right + @
      dup 0< if 2drop nip exit then
      nip swap 1+ swap
    else
      dup cells left + @
      dup 0< if 2drop nip exit then
      nip swap 1+ swap
    then
  again ;

: lcg ( n -- n' ) 1103515245 * 12345 + 2147483647 and ;

: build-tree ( -- )
  0 ncount !
  50000 insert
  12345
  N 1 do
    lcg dup N mod insert
  loop
  drop ;

: search-all ( -- sum )
  0
  12345
  N 1 do
    lcg dup N mod find-depth rot + swap
  loop
  drop ;

: main build-tree search-all . cr ;

File: compiler/bench/bstfind.c
// expected: 99999
#include <stdio.h>
#define N 100000

int left_child[N];
int right_child[N];
int key[N];
int node_count = 0;

void insert(int val) {
    if (node_count == 0) {
        key[0] = val;
        left_child[0] = -1;
        right_child[0] = -1;
        node_count = 1;
        return;
    }
    int curr = 0;
    while (1) {
        if (val < key[curr]) {
            if (left_child[curr] < 0) {
                left_child[curr] = node_count;
                key[node_count] = val;
                left_child[node_count] = -1;
                right_child[node_count] = -1;
                node_count++;
                return;
            }
            curr = left_child[curr];
        } else {
            if (right_child[curr] < 0) {
                right_child[curr] = node_count;
                key[node_count] = val;
                left_child[node_count] = -1;
                right_child[node_count] = -1;
                node_count++;
                return;
            }
            curr = right_child[curr];
        }
    }
}

int find_depth(int val) {
    int curr = 0;
    int depth = 0;
    while (curr >= 0) {
        if (key[curr] == val) return depth;
        if (val < key[curr]) {
            curr = left_child[curr];
        } else {
            curr = right_child[curr];
        }
        depth++;
    }
    return depth;
}

int lcg(int n) {
    return (n * 1103515245 + 12345) & 2147483647;
}

void build_tree() {
    node_count = 0;
    insert(50000);
    int seed = 12345;
    for (int i = 1; i < N; i++) {
        seed = lcg(seed);
        insert(seed % N);
    }
}

long search_all() {
    long sum = 0;
    int seed = 12345;
    for (int i = 1; i < N; i++) {
        seed = lcg(seed);
        sum += find_depth(seed % N);
    }
    return sum;
}

int main() {
    build_tree();
    printf("%ld\n", search_all());
    return 0;
}
